subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AnnotationProcessorDetectionIntegrationTest.groovy

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2013 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.java.compile.incremental
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.CompilationOutputsFixture
+import org.gradle.language.fixtures.IncrementalAnnotationProcessorFixture
+import org.gradle.language.fixtures.NonIncrementalAnnotationProcessorFixture
+
+class AnnotationProcessorDetectionIntegrationTest extends AbstractIntegrationSpec {
```

Comment:
Missing test cases:

- Starting with a non-incremental processor, do a compilation, then remove the non-incremental processor. The following compilation should be a full one to clean up its old outputs
- A processor using a later Incap feature that this version of Gradle can't support (e.g. aggregating processors) is treated as non-incremental
- An incap version with an unsupported mapping file format is on the processor path - compilation is non-incremental
- A processor or some bug in Incap lead to a corrupt mapping file. The next compilation is a full one.
- A processor has a malformed Incap meta-inf file. It is treated like a non-incremental one.
- It would be nice to show the processor working on other things than just the top-level type. E.g. on a method instead, to showcase that the right source dependencies are still added.
- The generated file is deleted - Gradle recompiles the source file
- The source file is incrementally changed, the target file has the right contents. The current processors' output is completely independent of the contents of the source file. It would be nice to do something like 1 target method for every source method or similar to make sure that on each incremental change we actually overwrite the target. 
- An unrelated Java file is changed - the generated sources are untouched

Reply:
I added all these, although one of them (mapping file) is still WIP.